File: backend/src/main/java/com/tutict/finalassignmentbackend/entity/DeductionRecord.java
package com.tutict.finalassignmentbackend.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;

import java.io.Serial;
import java.io.Serializable;
import java.time.LocalDateTime;

/**
 * 扣分记录表实体类，对应数据库表 "deduction_record"
 * 记录驾驶员的扣分信息
 */
@Data
@TableName("deduction_record")
public class DeductionRecord implements Serializable {
    /**
     * 序列化版本 UID
     */
    @Serial
    private static final long serialVersionUID = 1L;

    /**
     * 扣分记录 ID，主键，自动增长
     */
    @TableId(value = "deduction_id", type = IdType.AUTO)
    private Long deductionId;

    /**
     * 违法记录 ID
     */
    @TableField("offense_id")
    private Long offenseId;

    /**
     * 驾驶员 ID
     */
    @TableField("driver_id")
    private Long driverId;

    /**
     * 扣分分值
     */
    @TableField("deducted_points")
    private Integer deductedPoints;

    /**
     * 扣分时间
     */
    @TableField("deduction_time")
    private LocalDateTime deductionTime;

    /**
     * 记分周期 (如: 2025-01-01至2026-01-01)
     */
    @TableField("scoring_cycle")
    private String scoringCycle;

    /**
     * 处理人
     */
    @TableField("handler")
    private String handler;

    /**
     * 处理部门
     */
    @TableField("handler_dept")
    private String handlerDept;

    /**
     * 审批人
     */
    @TableField("approver")
    private String approver;

    /**
     * 审批时间
     */
    @TableField("approval_time")
    private LocalDateTime approvalTime;

    /**
     * 状态 (Effective, Cancelled, Restored)
     */
    @TableField("status")
    private String status;

    /**
     * 恢复时间
     */
    @TableField("restore_time")
    private LocalDateTime restoreTime;

    /**
     * 恢复原因
     */
    @TableField("restore_reason")
    private String restoreReason;

    /**
     * 创建时间
     */
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /**
     * 更新时间
     */
    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    /**
     * 创建人
     */
    @TableField("created_by")
    private String createdBy;

    /**
     * 更新人
     */
    @TableField("updated_by")
    private String updatedBy;

    /**
     * 软删除时间戳
     */
    @TableField("deleted_at")
    @TableLogic(value = "null", delval = "now()")
    private LocalDateTime deletedAt;

    /**
     * 备注
     */
    @TableField("remarks")
    private String remarks;
}

File: flutter_app/lib/features/model/deduction_record.dart
class DeductionRecordModel {
  final int? deductionId;
  final int? offenseId;
  final int? driverId;
  final int? deductedPoints;
  final DateTime? deductionTime;
  final String? scoringCycle;
  final String? handler;
  final String? handlerDept;
  final String? approver;
  final DateTime? approvalTime;
  final String? status;
  final DateTime? restoreTime;
  final String? restoreReason;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? remarks;

  const DeductionRecordModel({
    this.deductionId,
    this.offenseId,
    this.driverId,
    this.deductedPoints,
    this.deductionTime,
    this.scoringCycle,
    this.handler,
    this.handlerDept,
    this.approver,
    this.approvalTime,
    this.status,
    this.restoreTime,
    this.restoreReason,
    this.createdAt,
    this.updatedAt,
    this.remarks,
  });

  DeductionRecordModel copyWith({
    int? deductionId,
    int? offenseId,
    int? driverId,
    int? deductedPoints,
    DateTime? deductionTime,
    String? scoringCycle,
    String? handler,
    String? handlerDept,
    String? approver,
    DateTime? approvalTime,
    String? status,
    DateTime? restoreTime,
    String? restoreReason,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? remarks,
  }) {
    return DeductionRecordModel(
      deductionId: deductionId ?? this.deductionId,
      offenseId: offenseId ?? this.offenseId,
      driverId: driverId ?? this.driverId,
      deductedPoints: deductedPoints ?? this.deductedPoints,
      deductionTime: deductionTime ?? this.deductionTime,
      scoringCycle: scoringCycle ?? this.scoringCycle,
      handler: handler ?? this.handler,
      handlerDept: handlerDept ?? this.handlerDept,
      approver: approver ?? this.approver,
      approvalTime: approvalTime ?? this.approvalTime,
      status: status ?? this.status,
      restoreTime: restoreTime ?? this.restoreTime,
      restoreReason: restoreReason ?? this.restoreReason,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      remarks: remarks ?? this.remarks,
    );
  }

  factory DeductionRecordModel.fromJson(Map<String, dynamic> json) {
    return DeductionRecordModel(
      deductionId: json['deductionId'],
      offenseId: json['offenseId'],
      driverId: json['driverId'],
      deductedPoints: json['deductedPoints'],
      deductionTime: json['deductionTime'] != null
          ? DateTime.tryParse(json['deductionTime'])
          : null,
      scoringCycle: json['scoringCycle'],
      handler: json['handler'],
      handlerDept: json['handlerDept'],
      approver: json['approver'],
      approvalTime: json['approvalTime'] != null
          ? DateTime.tryParse(json['approvalTime'])
          : null,
      status: json['status'],
      restoreTime: json['restoreTime'] != null
          ? DateTime.tryParse(json['restoreTime'])
          : null,
      restoreReason: json['restoreReason'],
      createdAt:
          json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
      updatedAt:
          json['updatedAt'] != null ? DateTime.tryParse(json['updatedAt']) : null,
      remarks: json['remarks'],
    );
  }

  Map<String, dynamic> toJson() => {
        'deductionId': deductionId,
        'offenseId': offenseId,
        'driverId': driverId,
        'deductedPoints': deductedPoints,
        'deductionTime': deductionTime?.toIso8601String(),
        'scoringCycle': scoringCycle,
        'handler': handler,
        'handlerDept': handlerDept,
        'approver': approver,
        'approvalTime': approvalTime?.toIso8601String(),
        'status': status,
        'restoreTime': restoreTime?.toIso8601String(),
        'restoreReason': restoreReason,
        'createdAt': createdAt?.toIso8601String(),
        'updatedAt': updatedAt?.toIso8601String(),
        'remarks': remarks,
      };
}
